http-client/Data/KeyedPool.hs

```
@@ -0,0 +1,316 @@
+        join $ atomically $ do
+            m <- readTVar var
+            case m of
+                PoolClosed -> return (return ())
+                PoolOpen idleCount m
+                    | Map.null m -> retry
+                    | otherwise -> do
+                        (m', toDestroy) <- findStale idleCount m
+                        writeTVar var m'
+                        return $ do
+                            mask_ (mapM_ safeDestroy toDestroy)
+                            loop
+
+    findStale :: Int
+              -> Map key (PoolList resource)
+              -> STM (PoolMap key resource, [resource])
+    findStale idleCount m = do
+        -- We want to make sure to get the time _after_ any delays
+        -- occur due to the retry call above. Since getCurrentTime has
+        -- no side effects outside of the STM block, this is a safe
+        -- usage.
+        now <- unsafeIOToSTM getCurrentTime
+        let isNotStale time = 30 `addUTCTime` time >= now
+        let findStale' toKeep toDestroy [] =
+                (Map.fromList (toKeep []), toDestroy [])
+            findStale' toKeep toDestroy ((key, plist):rest) =
+                findStale' toKeep' toDestroy' rest
+              where
+                -- Note: By definition, the timestamps must be in
+                -- descending order, so we don't need to traverse the
+                -- whole list.
+                (notStale, stale) = span (isNotStale . fst) $ plistToList plist
+                toDestroy' = toDestroy . (map snd stale++)
+                toKeep' =
+                    case plistFromList notStale of
+                        Nothing -> toKeep
+                        Just x -> toKeep . ((key, x):)
+        let (toKeep, toDestroy) = findStale' id id (Map.toList m)
+        let idleCount' = idleCount - length toDestroy
+        return (PoolOpen idleCount' toKeep, toDestroy)
+
+    safeDestroy x = destroy x `catch` \(_ :: SomeException) -> return ()
```

I don't know if we actually want to silently catch all exceptions here. We should at least somehow log them because it's very important to know if destroying a resource didn't work for some reason.